swig/cpp/src/Session.cpp

```diff
@@ -0,0 +1,281 @@
+    case(SR_ERR_DISCONNECT):
+        throw runtime_error(sr_strerror(SR_ERR_DISCONNECT));
+    case(SR_ERR_MALFORMED_MSG):
+        throw runtime_error(sr_strerror(SR_ERR_MALFORMED_MSG));
+    case(SR_ERR_UNSUPPORTED):
+        throw runtime_error(sr_strerror(SR_ERR_UNSUPPORTED));
+    case(SR_ERR_UNKNOWN_MODEL):
+        throw runtime_error(sr_strerror(SR_ERR_UNKNOWN_MODEL));
+    case(SR_ERR_BAD_ELEMENT):
+        throw runtime_error(sr_strerror(SR_ERR_BAD_ELEMENT));
+    case(SR_ERR_VALIDATION_FAILED):
+        throw runtime_error(sr_strerror(SR_ERR_VALIDATION_FAILED));
+    case(SR_ERR_COMMIT_FAILED):
+        throw runtime_error(sr_strerror(SR_ERR_COMMIT_FAILED));
+    case(SR_ERR_DATA_EXISTS):
+        throw runtime_error(sr_strerror(SR_ERR_DATA_EXISTS));
+    case(SR_ERR_DATA_MISSING):
+        throw runtime_error(sr_strerror(SR_ERR_DATA_MISSING));
+    case(SR_ERR_UNAUTHORIZED):
+        throw runtime_error(sr_strerror(SR_ERR_UNAUTHORIZED));
+    case(SR_ERR_LOCKED):
+        throw runtime_error(sr_strerror(SR_ERR_LOCKED));
+    case(SR_ERR_TIME_OUT):
+        throw runtime_error(sr_strerror(SR_ERR_TIME_OUT));
+    }
+}
+
+Session::Session(Connection& sys, const sr_conn_options_t opts, const char *user_name)
+{
+    int rc;
+    _opts = opts;
+    _datastore = SR_DS_STARTUP;
+
+    if (user_name == NULL) {
+        /* start session */
+        rc = sr_session_start(sys.get_conn(), _datastore, _opts, &_sess);
+        if (SR_ERR_OK != rc) {
+            goto cleanup;
+        }
+    } else {
+        /* start session */
+        rc = sr_session_start_user(sys.get_conn(), user_name, _datastore, _opts, &_sess);
+        if (SR_ERR_OK != rc) {
+            goto cleanup;
+        }
+    }
+
+    return;
+
+cleanup:
+    throw_excption(SR_ERR_INIT_FAILED);
+    //TODO error handling
+    return;
+}
+
+int Session::session_stop()
+{
+    int ret = sr_session_stop(_sess);
+    if (ret != SR_ERR_OK) {
+        throw_excption(ret);
+    }
+    return ret;
+}
+
+int Session::get_last_error(Error& err)
+{
+    int ret = sr_get_last_error(_sess, &err.info);
+    if (ret != SR_ERR_OK) {
+        throw_excption(ret);
+    }
+    return ret;
+}
+
+int Session::get_last_errors(Error& err)
+{
+    int ret = sr_get_last_errors(_sess, &err.info, &err.cnt);
+    if (ret != SR_ERR_OK) {
+        throw_excption(ret);
+    }
+    return ret;
+}
+
+int Session::list_schemas(Schema& schema)
+{
+    int ret = sr_list_schemas(_sess, &schema.sch, &schema.cnt);
+    if (ret != SR_ERR_OK) {
+        throw_excption(ret);
+    }
+    return ret;
+}
+
+int Session::get_schema(Schema& schema, const char *module_name, const char *revision,
+	       	const char *submodule_name,  sr_schema_format_t format)
+{
+    int ret = sr_get_schema(_sess, module_name, revision, submodule_name, format, &schema.content);
+    if (ret != SR_ERR_OK) {
+        throw_excption(ret);
+    }
+    return ret;
+}
+
+int Session::get_item(const char *xpath, Value* value)
```

Comment:
Why has this method an argument of type "pointer to the Value" while the others have "a reference to Value"?